model/src/id/mod.rs

```diff
@@ -0,0 +1,711 @@
+            }
+
+            fn visit_newtype_struct<D: Deserializer<'de>>(
+                self,
+                deserializer: D,
+            ) -> Result<Self::Value, D::Error> {
+                deserializer.deserialize_any(IdVisitor::new())
+            }
+
+            fn visit_str<E: DeError>(self, v: &str) -> Result<Self::Value, E> {
+                let value = v.parse().map_err(|_| {
+                    let unexpected = Unexpected::Str(v);
+
+                    DeError::invalid_value(unexpected, &"non zero u64 string")
+                })?;
+
+                self.visit_u64(value)
+            }
+        }
+
+        deserializer.deserialize_any(IdVisitor::new())
+    }
+}
+
+impl<T> Display for Id<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        Display::fmt(&self.value.get(), f)
+    }
+}
+
+impl<T> From<NonZeroU64> for Id<T> {
+    fn from(id: NonZeroU64) -> Self {
+        Self::from_nonzero(id)
+    }
+}
+
+impl<T> FromStr for Id<T> {
+    type Err = ParseIntError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        NonZeroU64::from_str(s).map(Self::from_nonzero)
+    }
+}
+
+impl<T> Eq for Id<T> {}
+
+impl<T> Hash for Id<T> {
+    fn hash<U: Hasher>(&self, state: &mut U) {
+        state.write_u64(self.value.get());
+    }
+}
+
+impl<T> Ord for Id<T> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.value.cmp(&other.value)
+    }
+}
+
+impl<T> PartialEq for Id<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.value == other.value
+    }
+}
+
+impl<T> PartialEq<i64> for Id<T> {
+    fn eq(&self, other: &i64) -> bool {
+        u64::try_from(*other)
+            .map(|v| v == self.value.get())
+            .unwrap_or_default()
+    }
+}
+
+impl<T> PartialEq<Id<T>> for i64 {
+    fn eq(&self, other: &Id<T>) -> bool {
+        u64::try_from(*self)
+            .map(|v| v == other.value.get())
+            .unwrap_or_default()
+    }
+}
+
+impl<T> PartialEq<u64> for Id<T> {
+    fn eq(&self, other: &u64) -> bool {
+        self.value.get() == *other
+    }
+}
+
+impl<T> PartialEq<Id<T>> for u64 {
+    fn eq(&self, other: &Id<T>) -> bool {
+        other.value.get() == *self
+    }
+}
+
+impl<T> PartialOrd for Id<T> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        self.value.partial_cmp(&other.value)
+    }
+}
+
+impl<T> Serialize for Id<T> {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        // Avoid requiring a Copy trait bound by simply reconstructing self.
+        let copy = Self::from_nonzero(self.value);
+        let formatter = IdStringDisplay::new(copy);
+
+        serializer.serialize_newtype_struct("Id", &formatter)
+    }
+}
+
+impl<T> TryFrom<i64> for Id<T> {
+    type Error = TryFromIntError;
+
+    fn try_from(value: i64) -> Result<Self, Self::Error> {
+        let signed_nonzero = NonZeroI64::try_from(value)?;
+        let unsigned_nonzero = NonZeroU64::try_from(signed_nonzero)?;
```

Comment:
This would error for negative `value`s, and I want to check if that is intentional.
If it is intentional, what is this impl useful for? IDs generally wouldn't fit into only the positive part of an `i64`, so I don't see why you'd store it that way if you can't use the negative range.
Same question applies to the `PartialEq<i64>` impl.

Reply:
Ids can fit in `i64`.
The `i64` max is: `9223372036854775807` and for example this is the id of the twilight server: `745809834183753828`.
As you can see ids easily fit into the `i64` max.

Reply:
Yes, this is correct. IDs can not be negative. This PR includes support for comparing with and creating from i64s because many databases do not support unsigned 64-bit integers.